### PR TITLE
modification of kc.sh to remove param eval

### LINKED ISSUE #22337: kc.sh errors if using characters like semicolon inside the arguments
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

dist/quarkus

### Describe the bug

There are errors when using the `kc.sh` script if the parameters contain problematic characters like semicolon `;`. One easy example is the mssql `db-url`:

```bash
./kc.sh start --db=mssql --hostname=localhost --db-username=keycloak "--db-url=jdbc:sqlserver://127.0.0.1:1433;DatabaseName=keycloak;integratedSecurity=false;trustServerCertificate=true;encrypt=No;sendStringParametersAsUnicode=false" "--db-password=Keycloak01!" --transaction-xa-enabled=false
Changes detected in configuration. Updating the server image.
Updating the configuration and installing your custom providers, if any. Please wait.
2023-08-09 10:47:01,142 INFO  [io.quarkus.deployment.QuarkusAugmentor] (main) Quarkus augmentation completed in 2988ms
Server configuration updated and persisted. Run the following command to review the configuration:

	kc.sh show-config

Next time you run the server, just run:

	kc.sh start --optimized --hostname=localhost --db-username=keycloak --db-url=jdbc:sqlserver://127.0.0.1:1433

./kc.sh: line 128: --db-password=Keycloak01!: command not found
```

If you see the semicolon is breaking the line and the rest of parameters are not taken into account. I think we need to quote `${VAR@Q}` the arguments in the script. I suppose there can be other errors with passwords and so on if they contain problematic chars like `$"` or similar.

As a workaround you can use the `keycloak.conf` of env variables to pass the options.

### Version

22.0.1

### Expected behavior

The `kc.sh` should work OK and the chars should be escaped/quoted.

### Actual behavior

Errors in `kc.sh`.

### How to Reproduce?

See the example execution in the description.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shawkins**
Removes the direct eval of parameters and env variables.

Closes #22337

### CODE REVIEW

**Comment by shawkins** [comment]
Based upon our internal discussion, removed the second commit that introduced a backwards compatibility flag.  Will fix up the test failures to get this out of draft.

**Comment by shawkins** [comment]
@mabartos it looks like this will require some refactoring of test classes like StoreProvider.  It is putting some of the arguments in single quotes, which will work only if they are evaled.  It will probably need to individually add the option and the value instead.  Do you have any objection to that change?

**Comment by mabartos** [comment]
> @mabartos it looks like this will require some refactoring of test classes like StoreProvider. It is putting some of the arguments in single quotes, which will work only if they are evaled. It will probably need to individually add the option and the value instead. Do you have any objection to that change?

I don't have any objection. Go ahead with that, it's one of the possible supported approaches for CLI.

**Comment by rmartinc** [comment]
Juat one comment, is this working with a space? For example example in the password, something line `--db-password "pass with space"`.

**Comment by shawkins** [comment]
> Juat one comment, is this working with a space? For example example in the password, something line --db-password "pass with space"

Assuming you mean as shell arguments, then "pass with space" will be seen as a single argument without quotes.

**Comment by rmartinc** [comment]
I did a little test with mysql and I think spaces are not admitted now (which is expected AFAIK). For example:

```
./kc.sh start-dev --db mysql --db-username keycloak --db-url "jdbc:mysql://localhost:3306/keycloak" --db-password "keycloak 01"
Option '--db-password' expects a single value (password)
Try 'kc.sh start-dev --help' for more information on the available options.
```

I don't know if spaces are necessary or not, but with this fix I think they are not anymore. So just take care.

**Comment by shawkins** [comment]
> I don't know if spaces are necessary or not, but with this fix I think they are not anymore. So just take care.

This has to do with the intermediate accumulation logic for CONFIG_ARGS, I'll address that.

**Comment by rmartinc** [comment]
@shawkins This is more or less what I did in this [branch](https://github.com/rmartinc/keycloak/commit/043650fe81933dbe48950c3e3c047abbf215185a) for the closed issue. The issue in the upgrade test is because we are using both versions (old and new) and the double quote was used before and now it's not ok. I did a direct fix in that branch, but maybe we can check the version we are upgrading from, and modify to add the quotes only in that case.

I think this is a bug so I'm OK with the change, but just note we are changing the double quote interpretation and, if someone is using it right now, it won't work after this PR. So we need a note in the release notes if we decide to continue with the change.

Just one more note, I would also include the changes I did for the windows bat script if we continue with the PR. Windows also has issues with some characters like `&`.

**Comment by vmuzikar** [comment]
Is Win (`kc.bat`) also affected by this? We should have a unified behaviour.

**Comment by rmartinc** [comment]
> Do you want me to directly do that, or do you want to push a change into this branch?
> Is Win (kc.bat) also affected by this? We should have a unified behaviour.

Windows is not affected by the same double quoting behavior. But it's affected by using some weird characters in the parameters (like `&`, used in mariadb url for example, if I remember correctly). Maybe we can work on windows changes in a followup issue better.

In general I think this PR is OK, it's very similar to what I did before. And I'm also in favor of not adding a compatibility option. If you need to change your script to add the option, you can directly change it to make it work properly (so I don't see any benefit).

**Comment by shawkins** [comment]
Based upon feedback from @abstractj direct evaluation of parameters or environment variables is considered a security weakness and should be avoided.  So we should choose one of the remaining options:

- expect xargs
- or make a simplifying assumption about what the env variables can contain

**Comment by vmuzikar** [comment]
For the maintainability sake, I think the way to go is with xargs to avoid introducing yet additional complexity to the script. Looking at [findutils dependencies](http://rpmfind.net/linux/RPM/fedora/devel/rawhide/x86_64/f/findutils-4.9.0-5.fc39.x86_64.html), it does not look so bad to me.

@ASzc @abstractj Are you comfortable with re-introducing findutils to the image?

**Comment by shawkins** [comment]
@vmuzikar without xargs the handling of the env variables converted into arguments could look like (in the pr currently):

```
old_ifs=$IFS
IFS=" "
for i in $JAVA_OPTS
do
  JAVA_RUN_OPTS="$JAVA_RUN_OPTS $(quote "$i")"
done
IFS=$old_ifs
```
Basically just split on space and quote each of the tokens.  This will break on any usage of quoting, escaping, or a value that contains a space.

With xargs it looks like:
```
args() {
  while IFS= read -r entry; do
    result="$result $(quote "$entry")"
  done
  echo $result
}

JAVA_RUN_OPTS=$(echo "$JAVA_OPTS" | xargs printf '%s\n' | args)
```
Quoting will work, as will spaces: -X="some option" or -X=some\ option
However a hard newline is problematic.  If quoted it causes xargs to error out because of an unmatched quote.  And the only escaping that is supported is for non-newline blanks.

**Comment by ASzc** [comment]
@vmuzikar I checked the RPM list from a couple builds, and if we modify the server Dockerfile as follows, the only extra package we will get is findutils (no new dependencies). Also, findutils almost never gets CVEs itself. So, I don't have an issue adding findutils.

The bash wizard part of me wants to dig in an find a way for this to work without xargs, but I don't have the time for that! The extra package is an ok solution.

```diff
diff --git a/quarkus/container/Dockerfile b/quarkus/container/Dockerfile
index c40db77e2d..5a98733bf9 100644
--- a/quarkus/container/Dockerfile
+++ b/quarkus/container/Dockerfile
@@ -17,7 +17,7 @@ RUN mv /tmp/keycloak/keycloak-* /opt/keycloak && mkdir -p /opt/keycloak/data
 RUN chmod -R g+rwX /opt/keycloak
 
 ADD ubi-null.sh /tmp/
-RUN bash /tmp/ubi-null.sh java-17-openjdk-headless glibc-langpack-en
+RUN bash /tmp/ubi-null.sh java-17-openjdk-headless glibc-langpack-en findutils
 
 FROM registry.access.redhat.com/ubi9-micro
 ENV LANG en_US.UTF-8
```

**Comment by vmuzikar** [comment]
@ASzc Thanks for confirming, xargs it is then.

> The bash wizard part of me wants to dig in an find a way for this to work without xargs

Yeah, we discussed with @shawkins it is definitely possible but agreed the xargs way is cleaner and therefore more easily maintainable. ;)

**Comment by shawkins** [comment]
@vmuzikar @ASzc this now includes xargs for handling env variables.  As noted above, the only really problematic character moving forward would be a hard newline in an env variable.  The docs can spell out more the expectation of env variables if needed.

This is layered on @rmartinc commit, so it includes the tweaks to the kc.bat as well along with the fix to the migration test that adds back in double quoting.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java [hunk 1]
```diff
@@ -176,6 +176,13 @@ protected List<String> configureArgs(List<String> args) {
     private ProcessBuilder getProcessBuilder() {
         Map<String, String> env = new HashMap<>();
         String[] processCommands = getArgs(env).toArray(new String[0]);
+        if (suiteContext.get().isAuthServerMigrationEnabled() && configuration.getImportFile() != null) {
+            for (int i = 0; i < processCommands.length; i++) {
+                if (processCommands[i].startsWith("--db-url=")) {
+                    processCommands[i]= "--db-url=\"" + processCommands[i].substring(9) + "\"";
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: I wonder if we're not hiding from the tests too much the fact that quotes need to be used.

**Review by vmuzikar — COMMENTED** [review]
@shawkins Thank you. In general LGTM but I'd like a green tick from @ASzc and @Pepo48 as well. ;)

File: docs/documentation/release_notes/topics/23_0_0.adoc [hunk 2]
```diff
@@ -13,3 +13,17 @@ the latest FAPI 2 draft specifications when communicating with your clients. Tha
...
+
+```
+bin/kc.sh start --db postgres --db-username keycloak --db-url "\"jdbc:postgresql://localhost:5432/keycloak?ssl=false&connectTimeout=30\"" --db-password keycloak --hostname localhost
+```
+
+Use a single escape:
+
+```
+bin/kc.sh start --db postgres --db-username keycloak --db-url "jdbc:postgresql://localhost:5432/keycloak?ssl=false&connectTimeout=30" --db-password keycloak --hostname localhost
+```
```
[the remark this answers is not in the document]
Comment (shawkins) [reply docs/documentation/release_notes/topics/23_0_0.adoc:29]: @vmuzikar do you want it in both, or just the upgrading guide?

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/documentation/release_notes/topics/23_0_0.adoc:29]: I think upgrading guide is enough.

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java (hunk 1 above)
Comment (shawkins) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: @vmuzikar do you want this to be version specific somehow - check if the legacy version specifically requires double quoting?

Comment (vmuzikar) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: > do you want this to be version specific

I'm not sure I follow. :) The testsuite expects only one Keycloak version.

Comment (shawkins) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: This is a migration test my understanding from https://github.com/keycloak/keycloak/pull/22585#issuecomment-1695668307 is that we're initially creating an older instance.

**Review by ASzc — APPROVED** [review]
Changes to `quarkus/container/Dockerfile` LGTM

File: testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java (hunk 1 above)
Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: @vmuzikar Yes, the migration tests first start an old version (IIRC it's 19 by default), then the current version to perform the upgrade, then execute the tests. The MSSQL db needs the quotes. So we are using both type of arguments in the same test (double quotes in 19, and no double quotes in current version). As @shawkins comments we can check the old version (<23 for example), but we need something similar to this. I wouldn't do something very complicated, if the migrations tests change we can also change this part.

Comment (vmuzikar) [reply testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakQuarkusServerDeployableContainer.java:182]: Ah, sorry, I missed the migration part. Makes sense then, thanks for clarifying.

**Comment by Pepo48** [comment]
> > Do you want me to directly do that, or do you want to push a change into this branch?
> > Is Win (kc.bat) also affected by this? We should have a unified behaviour.
> 
> Windows is not affected by the same double quoting behavior. But it's affected by using some weird characters in the parameters (like `&`, used in mariadb url for example, if I remember correctly). Maybe we can work on windows changes in a followup issue better.
> 
> In general I think this PR is OK, it's very similar to what I did before. And I'm also in favor of not adding a compatibility option. If you need to change your script to add the option, you can directly change it to make it work properly (so I don't see any benefit).

On top of that, even the `=`equal signs within the jdbc url string are problematic. Given `--db-url "jdbc:postgresql://localhost:5432/keycloak?ssl=false&connectTimeout=30"` as an example, it first processes `--db-url "jdbc:postgresql://localhost:5432/keycloak?ssl` and then it consumes `false` as a separate parameter. Thus the error message `Option '--db-url' expects a single value (jdbc-url)`.

With the way we parse the parameters it seems that it is not possible to preserve the equal signs as a part of parameter values. I could also reproduce the same behaviour with a shorter demo script that mimicked our way of parsing.

What worked for me is to use `%*` which is a special variable that contains all the passed parameters (represented by a list):
```
for /f "tokens=1*" %%x in ("%*") do (
    if /i "%%x"=="--db-url" ( :: the left part (parameter)
        set "DB_URL=%%y" :: the right part (value)
    )
)
```
 The question is whether it should be addressed in this PR or in a follow-up, as @rmartinc suggested.

TL;DR: parsing does not properly work on Windows when if jdbc url contains certain characters. Other than that, the proposed LGTM.

cc: @shawkins @vmuzikar.

File: quarkus/dist/src/main/content/bin/kc.bat
[the remark this answers is not in the document]
Comment (Pepo48) [reply quarkus/dist/src/main/content/bin/kc.bat:1]: @vmuzikar see my comment bellow.

**Comment by rmartinc** [comment]
@Pepo48 Have you tested with the changes in this PR? Because I tested with mssql and postgresql and I took note of the following startup and it was OK:

```
kc.bat start --hostname localhost --db mssql --db-username keycloak --db-url "jdbc:sqlserver://192.168.100.1:1433;DatabaseName=keycloak;integratedSecurity=false;trustServerCertificate=true;encrypt=No;sendStringParametersAsUnicode=false" --db-password "KiosKo_@01!" --transaction-xa-enabled false
```

Maybe there are still issues but at least a common boot command works (as commented I tested postgresql and mssql).

**Comment by Pepo48** [comment]
@rmartinc I tested it using the format mentioned in the release notes (the one that contains special characters). If it's not the case, then it's my bad. Sorry about that, I misunderstood it.

**Review by Pepo48 — APPROVED** [review]

**Comment by shawkins** [comment]
> I tested it using the format mentioned in the release notes (the one that contains special characters). If it's not the case, then it's my bad. Sorry about that, I misunderstood it.

@Pepo48 the release note is specific to kc.sh.  @rmartinc Were people using similar workarounds for kc.bat that need to be called out?

File: quarkus/dist/src/main/content/bin/kc.sh [hunk 3]
```diff
@@ -61,10 +64,16 @@ do
...
           ;;
       *)
+          OPT=$(esceval "$1")
           case "$1" in
             start-dev) CONFIG_ARGS="$CONFIG_ARGS --profile=dev $1";;
-            -D*) SERVER_OPTS="$SERVER_OPTS $1";;
-            *) CONFIG_ARGS="$CONFIG_ARGS $1";;
+            -D*) SERVER_OPTS="$SERVER_OPTS ${OPT}";;
+            *) case "$1" in
+                 --optimized | --help | --help-all | -h) PRE_BUILD=false;;
```
Comment (shawkins) [inline quarkus/dist/src/main/content/bin/kc.sh:72]: By the conclusion in #23071 we may not need to track --help, --help-all, nor -h in the shell scripts.  @mabartos WDYT?

Comment (vmuzikar) [reply quarkus/dist/src/main/content/bin/kc.sh:72]: @mabartos Can we get your input on this? :)

Comment (mabartos) [reply quarkus/dist/src/main/content/bin/kc.sh:72]: > By the conclusion in #23071 we may not need to track --help, --help-all, nor -h in the shell scripts, or for consistency should --show-config be added? @mabartos WDYT?

@shawkins @vmuzikar Yep, that's the question. Both variants might be a little bit controversial. It's true we do not need to track the `help` and `show-config` options in the script as they will not execute the augmentation phase due to this[1]. 

However, if they're not specified here, two processes are created. The first one, which is supposed to be for the 'pre-build', that will exit almost immediately, and then the second one, which is the application process itself. 

Comparison of both approaches:
1) Specify the options here
   **+** No additional process is created
   **-**  Worse maintainability

2) Do NOT specify the options here
   **-** Additional process is created
   **+**  Better maintainability

I'm kinda leaning towards option 2) for now, as these scripts are already quite messy. As all the `help` and `show-config` are terminal options, we do not need to care much about the performance. As I've mentioned, the process is quickly killed and does not consume resources in some big way. Thus, I prefer maintainability in this case.

**Future consideration:**
I don't like our execution scripts due to multiple reasons. The first one is described above, and the other is that I think these scripts should only manage the system stuff, not any comprehensive app logic. I'd even prefer to get rid of checking the `build`, `--optimized`, `kc.config.build-and-exit`, `kc.config.built` options in these scripts and rather move the handling them to the app logic. 

But we can discuss details elsewhere as it's out of the scope of this PR.    

[1] https://github.com/keycloak/keycloak/blob/a7d082e390ca128a44ad9e6ec6f5b8d2bfacfd0b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java#L130

**Review by mabartos — APPROVED** [review]
The minor improvements I've suggested can be done in a follow-up task.

@shawkins LGTM.

**Review by vmuzikar — APPROVED** [review]

**Comment by mabartos** [comment]
Created follow up for the mentioned options: https://github.com/keycloak/keycloak/issues/23972
